omp.c

```diff
@@ -0,0 +1,149 @@
+//*****************************************************************/
+// Author: 20Goto10
+// Description: This is the sequential version of the matrix
+//  multiplcation problem. This abstracts the movement of the
+//  iteration over the matrices into a linked list of recursive
+//  functions over a notional linked list of functions. This
+//  linked list is modeled as a recursive call to requesting
+//  additional work allowing eventually terminating in a notional
+//  wait method. This allows the threading of this solution to
+//  any amount, as each notional linked list is seperate from
+//  eachother during runtime, but aren't seperate at compile-
+//  time.
+// Model Example:
+//  let R = RequestWork
+//  let G = GenerateWork (solves a slot then calls R)
+//  let T = GenerateBarrierWait (acts as a wait/terminator for the
+//      work when no more work exists.)
+//  let R -> (G|T)
+//  Call stack looks like: R() -> G() -> G() -> ... -> T()
+//  where G eventually calls an R which returns a T halting the
+//  program
+//*****************************************************************/
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <omp.h>					//ADDED header
+
+#define SIZE 3
+#define SEED 34
+#define HIGH 5
+
+//Typedefs
+typedef bool (*WorkFunction)();
+
+//Stubs
+WorkFunction requestWork();
+WorkFunction generateDoWork(int x, int y);
+WorkFunction generateTerminate();
+int getDotProduct(int row, int col);
+void fillMatrix(int matrix[SIZE][SIZE]);
+void prettyPrint(char* name, int matrix[SIZE][SIZE]);
+
+//Global State
+int matA[SIZE][SIZE];
+int matB[SIZE][SIZE];
+int matSolution[SIZE][SIZE];
+int currentX = -1; //makes the iteration below simpler. Could solve the off by one down there too though.
+int currentY = 0;
+
+//entry
+int main()
+{  
+    srand(SEED);
+    fillMatrix(matA);
+    fillMatrix(matB);
+    
+    prettyPrint("A", matA);
+    prettyPrint("B", matB);
+    
+    //this is where the multi threading magic goes. Any amount of threads can make this call and have this work.
+    #pragma omp parallel			//ADDED multi threading magic
+    while(requestWork()());
+    
+    prettyPrint("Solution", matSolution);
+}
+
+//Returns lambda for next unit of work or a terminator.
+WorkFunction requestWork()
+{
+	#pragma omp critical			//ADDED critical section
+	{								//ADDED scoping
+		currentX++;
+		if(currentX >= SIZE) //handles wrap to next line
+		{
+			currentX = 0;
+			currentY++;
+		}
+	}								//ADDED scoping
+		
+    if(currentY >= SIZE) //handles early out
+    {
+        return generateTerminate();
+    }
+    
+    return generateDoWork(currentX, currentY);
```

Comment:
this is still a race condition. Notice in the pthread version we have this critical section encompass all checks to currentX and currentY, then we assign them to local variables inside this critical section, then pass those variables out to the return statements. This is because currentX and Y are shared memory, and can be changed prior to the return and subsequent currentY>= SIZE check, changing the values. I know this is a small one and would be tough to see in real life, but we should probably fix it.